test/test_integration.rb

```diff
@@ -389,7 +389,207 @@ def test_no_zombie_children
     worker_pids.each { |pid| Process.kill :TERM, pid }
     sleep 2
 
+    # should be empty if all old workers removed
+    waited_pids = worker_pids.map { |pid|
+      begin
+        Process.wait pid, Process::WNOHANG
+        pid
+      rescue Errno::ECHILD
+        nil # child is already reaped/removed
+      end
+    }.compact
+
     # Should return nil if Puma has correctly cleaned up
-    assert_nil Process.waitpid(-1, Process::WNOHANG)
+    assert_empty waited_pids
+  end
+
+  def test_worker_spawn_external_term
+    skip NO_FORK_MSG unless HAS_FORK
+
+    host = '127.0.0.1'
+    port = UniquePort.call
+    workers_booted = 0
+
+    conf = Puma::Configuration.new do |c|
+      c.bind "tcp://#{host}:#{port}"
+      c.threads 1, 1
+      c.workers 2
+      c.worker_shutdown_timeout 2
+      c.app TestApps::SLEEP
+      c.after_worker_fork { |idx| workers_booted += 1 }
```

Comment:
Why do we need a new mechanism for determining if the process has finished booting? We already have two methods.